Add create task job to automate task creation in corejobs

OUMIMANDAL opened the pull request:
This PR introduces a new job, CreateTaskJob, within the corejobs package to streamline the creation of tasks via the Openlane client. Features include:

Generic Tasks: Specify title and description; category and assignee are optional.

Policy Review Tasks: Automatically generates task title and description based on internal policy IDs and assigns a random user from the policy approver group.

Optional Delay: Schedule tasks to run at a future time.

Fully unit-tested using a mock Openlane client.

Job registration with Riverboat is not required; this PR focuses solely on defining the job and its functionality.

golanglemonade commented on corejobs/create_task_job.go:
```diff
@@ -0,0 +1,169 @@
+	log.Debug().Str("task_type", string(job.Args.Type)).Msg("starting task creation")
+
+	if w.olClient == nil {
+		cl, err := getOpenlaneClient(TaskConfig{}) // implement this to return your olClient
```
`getOpenlaneClient` takes in a `CustomDomainConfig`.  I'm going to push a PR to genericize this function so you can just use that and then you should be able to update this to pass in the config. Either way, you'll need to pass in the actual config, not an empty struct, e.g something like:

```
		cl, err := getOpenlaneClient(w.Config) 
```

The config will also need to be able to set those parameters as well.

golanglemonade commented on corejobs/create_task_job.go:
```diff
@@ -0,0 +1,169 @@
+		if err != nil {
+			return err
+		}
+		log.Info().Str("title", cfg.Title).Msg("Generic task created")
```
I would pull each of these out into their own function, e.g.

func (c * GenericConfig) genericTask() error {
    if c == nil {
       return fmt.Errorf("%w: missing GenericTaskConfig", ErrUnsupportedTask)
    }

    task := openlaneclient.CreateTaskInput{
	    Title:          cfg.Title,
	    Description:    cfg.Description,
	    Category:       cfg.Category,
	    OrganizationID: cfg.OrganizationID,
    }
  
    _, err := w.olClient.CreateTask(ctx, task)
    if err != nil {
	    return err
    }

	log.Info().Str("title", cfg.Title).Msg("Generic task created")

    return nil
}
```

This allows the work job to remain simple, even as the types expand

golanglemonade commented on corejobs/create_task_job.go:
```diff
@@ -0,0 +1,169 @@
+		assignee := policy.ApproversGroup.Users[rand.Intn(len(policy.ApproversGroup.Users))]
+
+		task := openlaneclient.CreateTaskInput{
+			Title:          fmt.Sprintf("Policy Review %s", cfg.InternalPolicyIDs[0]),
```
Use the name instead of the `id`
```suggestion
			Title:          fmt.Sprintf("Policy Review: %s", policy.Name),
```

golanglemonade commented on corejobs/create_task_job.go:
```diff
@@ -0,0 +1,169 @@
+		if cfg == nil || len(cfg.InternalPolicyIDs) == 0 {
+			return fmt.Errorf("%w: PolicyReview requires InternalPolicyIDs", ErrUnsupportedTask)
+		}
+
```
you'd want this in a loop, instead of doing the first id, e.g. :

```
for _, pID := range cfg.InternalPolicyIDs {
		policy, err := w.olClient.GetInternalPolicy(ctx, pID)
		if err != nil {
			return err
		}
... rest of the logic ...
}
```

OUMIMANDAL commented:
Hi @golanglemonade,

This PR only updates corejobs/create_task_job.go, corejobs/create_task_job_test.go, and mocks/mock_openlaneclient.go.
Other files listed (like go.mod/go.sum) are unrelated to my changes — please review on the core job updates and let me know any other changes i need to make 

Thanks again

golanglemonade commented:
> Hi @golanglemonade,
> 
> This PR only updates corejobs/create_task_job.go, corejobs/create_task_job_test.go, and mocks/mock_openlaneclient.go. Other files listed (like go.mod/go.sum) are unrelated to my changes — please review on the core job updates and let me know any other changes i need to make
> 
> Thanks again

Can you please rebase and remove extra files + fix any merge conflicts. It's hard to review this PR as is to see what was actually changed.

golanglemonade commented on h drop stash@{0}:
```diff
@@ -0,0 +1,2 @@
+stash@{0}: On add-create-task-job: temp unrelated changes
```
This file needs to be removed

golanglemonade commented on v:
```diff
@@ -0,0 +1,324 @@
+
```
Remove this file as well

golanglemonade commented on corejobs/create_task_job.go:
```diff
@@ -0,0 +1,119 @@
+		}
+
+		// Simulate a CreateTask or similar call
+		_, err := j.olClient.CloneBulkCSVControl(ctx, upload)
```
Same here; you shouldn't be cloning controls. It should be creating a task.

OUMIMANDAL commented:
Hi @golanglemonade,  
I’ve cleaned and rebased the PR — it now only includes the relevant `create_task_job.go` and test file.  
Kindly review when you have a moment. Thanks!

golanglemonade commented:
@OUMIMANDAL The logic here in the jobs are not correct - please see my open comments on why you have code to clone controls instead of creating a task object.

golanglemonade commented:
Closing this out as the PR doesn't address the underlying issue. Feel free to jump back in with a new PR if you would like, happy to answer any questions you have on the linked issue or in our discord channel.
